micro-rdk/src/common/ota.rs

```diff
@@ -334,71 +327,108 @@ impl<S: OtaMetadataStorage> OtaService<S> {
             uri = hyper::Uri::from_parts(parts).map_err(|e| OtaError::Other(e.to_string()))?;
         };
 
-        let mut num_tries = 0;
-        let (mut sender, conn) = loop {
-            num_tries += 1;
-            if num_tries == NUM_RETRY_CONN + 1 {
-                return Err(OtaError::Other(
-                    "failed to establish connection".to_string(),
-                ));
-            }
-            match self.connector.connect_to(&uri) {
-                Ok(connection) => {
-                    match connection.await {
-                        Ok(io) => {
-                            match http2::Builder::new(self.exec.clone())
-                                .max_frame_size(16_384) // lowest configurable
-                                .timer(H2Timer)
-                                .handshake(io)
-                                .await
-                            {
-                                Ok(pair) => break pair,
-                                Err(e) => {
-                                    log::error!("failed to build http request: {}", e);
+        Ok(uri)
+    }
+
+    /// Attempts to perform an OTA update.
+    /// On success, returns an `Ok(true)` or `Ok(false)` indicating if a reboot is necessary.
+    pub(crate) async fn update(&mut self) -> Result<bool, OtaError<S>> {
+        if !(self.needs_update()) {
+            return Ok(false);
+        }
+
+        let mut uri = self.parse_uri(&self.url)?;
+        let mut response: hyper::Response<hyper::body::Incoming>;
+
+        let conn = loop {
+            let (mut sender, conn) = loop {
```

Comment:
Having looked through the whole thing, I wonder if you really need this double loop and could just let `num_tries` drive the whole thing, especially as a way out from a redirect loop.

Reply:
I could have `num_tries` drive it. or a combination of `num_tries` and `num_redirects` if we wanted to go finer, but I'm ok with `num_tries`